cmd/dcrtime_checker/dcrtime_checker.go

```diff
@@ -0,0 +1,227 @@
+	h := sha256.New()
+	f, err := os.Open(filename)
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	if _, err = io.Copy(h, f); err != nil {
+		return "", err
+	}
+
+	return hex.EncodeToString(h.Sum(nil)), nil
+}
+
+func verifyAnchor(url, tx string, mr []byte) error {
+	u := url + tx + "/out"
+	r, err := http.Get(u)
+	if err != nil {
+		return fmt.Errorf("HTTP Get: %v", err)
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != http.StatusOK {
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			return fmt.Errorf("invalid body: %v %v",
+				r.StatusCode, body)
+		}
+		return fmt.Errorf("invalid dcrdata answer: %v %s",
+			r.StatusCode, body)
+	}
+
+	var txOuts []types.TxOut
+	d := json.NewDecoder(r.Body)
+	if err := d.Decode(&txOuts); err != nil {
+		return err
+	}
+
+	var done bool
+	for _, v := range txOuts {
+		if !types.IsNullDataScript(v.ScriptPubKeyDecoded.Type) {
+			continue
+		}
+		script, err := hex.DecodeString(v.ScriptPubKeyDecoded.Hex)
+		if err != nil {
+			return fmt.Errorf("invalid dcrdata script: %v", err)
+		}
+		data, err := txscript.PushedData(script)
+		if err != nil {
+			return fmt.Errorf("invalid script: %v", err)
+		}
+		if !bytes.Equal(data[0], mr) {
+			continue
+		}
+
+		// Everything is cool so mark it and break out
+		done = true
+		break
+	}
+	if !done {
+		return fmt.Errorf("merkle root not found: tx %v merkle %x",
+			tx, mr)
+	}
+
+	return nil
+}
+
+func _main() error {
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "dcrtime_checker [-h {dcrdatahost}|"+
+			"-testnet|-v] -f {file} -p {proof}\n\n")
+		flag.PrintDefaults()
+	}
+	flag.Parse()
+
+	// require -f
+	if *file == "" {
+		return fmt.Errorf("must provide -f")
+	}
+
+	// require -p
+	if *proof == "" {
+		return fmt.Errorf("must provide -p")
+	}
+
+	// Ensure proof looks correct
+	fProof, err := os.Open(*proof)
+	if err != nil {
+		return err
+	}
+	var vr v1.VerifyReply
+	decoder := json.NewDecoder(fProof)
+	if err := decoder.Decode(&vr); err != nil {
+		return fmt.Errorf("Could node decode VerifyReply: %v", err)
+	}
+
+	// Handle dcrtime host
+	if *dcrdataHost == "" {
+		if *testnet {
+			*dcrdataHost = "https://testnet.dcrdata.org/api/tx/"
+		} else {
+			*dcrdataHost = "https://explorer.dcrdata.org/api/tx/"
+		}
+	} else {
+		if !strings.HasSuffix(*dcrdataHost, "/") {
+			*dcrdataHost += "/"
+		}
+	}
+	// Get file digest
+	d, err := digestFile(*file)
+	if err != nil {
+		return err
+	}
+
+	if *verbose {
+		fmt.Printf("%v  %v\n", d, *file)
+	}
+
+	// Ensure file digest exists in the proof and that the saved answer was
+	// correct
+	var root *[32]byte
+	found := -1
+	for k, v := range vr.Digests {
+		if v.Digest != d {
+			continue
+		}
+
+		// Verify result of matching digest
+		if _, ok := v1.Result[v.Result]; !ok {
+			return fmt.Errorf("%v invalid error code %v\n", v.Digest,
+				v.Result)
+		}
+
+		// Verify merkle path.
+		root, err = merkle.VerifyAuthPath(&v.ChainInformation.MerklePath)
```

Comment:
I think we should have a step before to check the provided digest is part of the Merkle path, right? 
Otherwise, the proof can contain a valid Merkle path which doesn't have digest included in it.

Reply:
We do that. Read the comment right on top of the for loop on line 160.